### PR TITLE
[fix #447] Implement brand themes

### LINKED ISSUE #447: (9/26) Implement theming
In order to manage application of Protocol as it applies to Mozilla and Firefox design systems we will need to create a set of theming conventions that cascade down from a top level declarations.

## Requirements
* typographic sizing, color, and font classes should be separate. This separation will allow for down-the-road flexibility for creating different colorways, type scales, and font-usage patterns.
* in order to close this issue we should implement the following:

1. a class that implements the new firefox font stack and default document color scheme w/in a document
2. a class that implements the mozilla font stack and default document color scheme w/in a document
3. a class that implements the firefox type scale w/in a document
4. a class that implements the mozilla type scale w/in a document

## Recommendations

right now, there's an `mzp-t-firefox` class used to set Open Sans...i think we need to be a little more granular about the naming system here, something like this might work: 
```
<body class="mzp-t-brand-firefox mzp-t-usage-presentation">
```

In this version any color and font stack selectors of can be descendants of the `mzp-t-brand-firefox` class, and all scaling selectors can be descendants of of the `mzp-t-usage-presentation` class.

The reason for separating these is because once Protocol is used for app UI, it is very likely that the overall scale will change, so i could easily see a variant here with `mzp-t-usage-application` that keeps the Firefox theme but wants different sizings (and, for example, no `:visited` color on links).

Additionally, this leaves room for another optional base class (`mzp-t-colorway-inverted` or similar) that could override a color theme.

### PR DESCRIPTION
**Pull request by craigcook**
## Description

This finally adds basic theming to Protocol. Themable styles are assigned to variables defined in maps, then the `get-theme()` function can swap to a different map based on the value of a single theme variable.

This also adds an alternative condensed type scale that works somewhat similarly, swapping to a different map of sizes based on another variable. 

- [x] I have documented this change in the design system.
- [x] I have recorded this change in `CHANGELOG.md`.

### Issue
#447 - Implement theming
#405 - Update Firefox theme text colors
#410 - Update type scales

### Testing
https://demo1--mozilla-protocol.netlify.app/

This is a pretty big PR that touches a lot of files, but a lot of that is just tidying up comments and tweaking some docs. The real meat of the theming is in `includes/_themes.scss` and `includes/_functions.scss`

### CODE REVIEW

File: src/assets/sass/protocol/_theme-config.scss
Comment (craigcook) [inline src/assets/sass/protocol/_theme-config.scss:5]: Note for reviewer: I put this file in the root because I want to discourage consumers from editing any other files, especially since they're subject to change on future updates. Otherwise we could have defined these variables at the top of _lib, but I really don't like the idea of people messing with that file.

And defining them in protocol.scss wouldn't really work because it's really _lib that needs to know these values (they have to be defined _before_ _lib is imported) but maybe there's a better way.

**Comment by stephaniehobson** [comment]
We should publish a point release to fix this: https://github.com/mozilla/protocol/issues/611 before merging the theme changes. I'll still try to start the review though.

File: src/assets/sass/protocol/_theme-config.scss
Comment (stephaniehobson) [reply src/assets/sass/protocol/_theme-config.scss:5]: I think this is the best way to do it 👍

File: src/assets/sass/protocol/includes/_functions.scss
Comment (stephaniehobson) [inline src/assets/sass/protocol/includes/_functions.scss:100]: Error handling here too?

File: src/assets/sass/protocol/base/elements/_document.scss
Comment (stephaniehobson) [inline src/assets/sass/protocol/base/elements/_document.scss:15]: get-theme?

File: src/assets/sass/protocol/_theme-config.scss
Comment (stephaniehobson) [reply src/assets/sass/protocol/_theme-config.scss:5]: Wait, can this file be ridden with a local copy? Or are they going to have to edit a file in the npm package?

Comment (craigcook) [reply src/assets/sass/protocol/_theme-config.scss:5]: Crap you're right... it should be overridden locally, which means I need to rethink this... maybe we don't need a config file at all and we can just set the defaults at the top of _lib after all. I'll work on it.

Comment (stephaniehobson) [reply src/assets/sass/protocol/_theme-config.scss:5]: Well, these are identified as "default" in the config so the values should be over-rideable by a later declaration. In bedrock could we do it in the same place we do these definitions?

```
$font-path: '/media/fonts';
$image-path: '/media/protocol/img';
```

Comment (craigcook) [reply src/assets/sass/protocol/_theme-config.scss:5]: Yeah, that's how it works, but I wrote out all these comments as if someone would modify their local copy... which you could totally do if you're just copying Protocol (like I was doing for all my testing), but when it's used as a package and this file lives in a node_modules folder nobody should be editing it and these comments are a bit misleading.

We do still have to declare _some_ default, but we can just put it right in _lib instead of in this separate file since we don't expect people to edit it, and I'll work on updating the comments and theme docs to explain how/where to change themes in a local file.

File: src/assets/sass/protocol/includes/_functions.scss
Comment (craigcook) [reply src/assets/sass/protocol/includes/_functions.scss:64]: Looks like Sass already throws an error if the argument is missing so this whole `@if not $scale` bit isn't really needed, we get error handling for free.

File: src/assets/sass/protocol/base/elements/_forms.scss
Comment (craigcook) [reply src/assets/sass/protocol/base/elements/_forms.scss:115]: I didn't find it in any of the form markup but there were plenty of now-redundant instances elsewhere that I've removed.

File: src/assets/sass/protocol/base/_elements.scss
Comment (craigcook) [reply src/assets/sass/protocol/base/_elements.scss:25]: Moved to a new _common.scss file

**Comment by craigcook** [comment]
Updated.

File: src/assets/sass/protocol/components/forms/_form.scss
Comment (craigcook) [reply src/assets/sass/protocol/components/forms/_form.scss:60]: Now that you mention it, do we want to rename this to `.mzp-c-form-title`?

**Comment by craigcook** [comment]
Temporarily putting this on hold because we have a few other small things in the pipe that we want to get into bedrock quickly, so it'll be better to merge them ahead of this PR so they don't get tied to a major release.

File: src/assets/sass/protocol/base/elements/_document.scss
Comment (stephaniehobson) [inline src/assets/sass/protocol/base/elements/_document.scss:23]: Declaring a foreground colour without a background colour has always made me uncomfortable. Last time we talked about it we agreed to remove it: https://github.com/mozilla/protocol/issues/515

If you're going to add the text colour change here it needs a background color declaration as well, and then we need to deal with the fallout from that.

I think this is a good time to do that actually since this is a breaking change.

If you'd like I can put together a separate PR with the mzp-t-dark changes to go live with the theme changes.

File: src/assets/sass/protocol/components/forms/_form.scss
Comment (stephaniehobson) [reply src/assets/sass/protocol/components/forms/_form.scss:60]: Yes.

File: src/assets/sass/protocol/base/elements/_document.scss
Comment (craigcook) [reply src/assets/sass/protocol/base/elements/_document.scss:23]: I'm removing this. We don't really have a proper global dark theme, it's just been done on a per-component basis so I think adding such a sweeping rule here does more harm than good. And once I remove this I'll double-check all the components with dark versions.

**Comment by craigcook** [comment]
Believe it or not, this is ready for review again (!).

**Comment by craigcook** [comment]
Once again marking Do Not Merge temporarily so we can get a small update published first (#638)

File: src/assets/sass/protocol/includes/_lib.scss
Comment (stephaniehobson) [inline src/assets/sass/protocol/includes/_lib.scss:48]: Do we not also need variables?

**Review by stephaniehobson — APPROVED** [review]
R+ 🎭 

I suggest we make it Mozilla out of the box instead of Firefox out of the box. It's not a blocker if you disagree but it's a decision we have to live with going forward so now's the time to consider it.

**Comment by craigcook** [comment]
Updated!

**Pull request merged**
